src/compiler/crystal/loader/msvc.cr

```diff
@@ -32,28 +32,53 @@ class Crystal::Loader
     end
   end
 
-  # Returns the list of DLLs imported from the libraries specified in the given
-  # linker arguments. Used by the compiler for delay-loaded DLL support.
-  def self.search_dlls(args : Array(String), *, search_paths : Array(String) = default_search_paths) : Set(String)
-    search_paths, libnames = parse_args(args, search_paths)
-    dlls = Set(String).new
+  struct SearchLibResult
+    getter library_paths = [] of String
+    getter remaining_args = [] of String
+    getter(not_found) { [] of String }
+
+    def not_found?
+      @not_found
+    end
+  end
+
+  # Extracts the command-line arguments from *args* that add libraries and
+  # expands them to their absolute paths. Returns a `SearchLibResult` with those
+  # expanded paths, plus unused arguments and libraries that were not found.
+  def self.search_libraries(args : Array(String), *, search_paths : Array(String) = default_search_paths, extra_suffix : String? = nil) : SearchLibResult
+    result = SearchLibResult.new
+    search_paths, libnames = parse_args(args, search_paths, remaining: result.remaining_args)
 
     libnames.each do |libname|
-      search_paths.each do |directory|
-        library_path = File.join(directory, library_filename(libname))
-        next unless File.file?(library_path)
+      if found_path = search_library(libname, search_paths, extra_suffix)
+        result.library_paths << found_path
+      else
+        result.not_found << libname
+      end
+    end
+
+    result
+  end
 
-        Crystal::System::LibraryArchive.imported_dlls(library_path).each do |dll|
-          dlls << dll unless dll.compare("kernel32.dll", case_insensitive: true).zero?
+  private def self.search_library(libname, search_paths, extra_suffix)
+    if ::Path::SEPARATORS.any? { |separator| libname.includes?(separator) }
+      libname = File.expand_path(libname)
```

Comment:
Any reason you're using `File.expand_path` instead of `.join`?
`libname` is just the stem of a filename, so any of the additional semantics of `expand_path` should not make any difference here and you get the same result with `File.join(Dir.current, libname)`.